### DISCUSSION TITLE
TensorFlow Implementation of SwinT-ChARM (Zhu et al. ICLR 2022)

### DISCUSSION THREAD

**Discussion by Nikolai10** [tensorflow/compression#151] [Show and tell]
Dear TFC-Community, 

I have recently been working on a TF implementation of SwinT-ChARM ([Zhu et al. ICLR 2022](https://openreview.net/pdf?id=IDwN6xjHnK8)):
https://github.com/Nikolai10/SwinT-ChARM.

SwinT-ChARM can be considered as an extension to [Minnen et al. ICIP 2020](https://arxiv.org/pdf/2007.08739.pdf), where the conventional nonlinear parts are replaced by Swin-transformer blocks. As a result, SwinT-ChARM achieves a better overall rate-distortion-computation trade-off, making it a promising direction for resource-constrained environments.

In my implementation I have taken some care to closely follow the naming conventions presented by [Minnen et al.](https://github.com/tensorflow/compression/blob/master/models/ms2020.py), without changing the overall structure. I have also included a Google Colab demo, which you can find [here](https://colab.research.google.com/drive/1V-2yrtyrFJLLHHibq6p_EHV4j5wxezIQ?usp=sharing).

I hope this work will be useful for further exploration and learning.

Kind regards, 
Nikolai

**Comment by minnend** [comment]
Cool, nice work!

I was impressed with the swint-charm paper (and appreciate their use of "CHARM" as an acronym for our entropy model, which I've now adopted). So it's great to see a TF implementation here.

It's clear at this point that the simple "four conv layers" transforms we've been using for analysis and synthesis hold back the RD performance of our models. That's hardly new -- (Cheng 2020) used a better transform and there are probably earlier references from [CLIC 2019](http://compression.cc) -- but the swint-charm paper seems like a major step forward in terms of RD for a fast model.

Based on (Zhu 2022), I'm now wondering how much the attention window pattern matters in the context of compression. There have been many papers looking at alternatives that evaluate on classification tasks ([cswin](https://arxiv.org/abs/2107.00652), [axial attention](https://arxiv.org/abs/1912.12180), [maxvit](https://arxiv.org/abs/2204.01697), etc.) but I don't know if their positive results transfer to dense prediction tasks in general and image/video compression specifically.

One interpretation of (Zhu 2022) is that transformers are key for the autoencoder in a compression model based on nonlinear transform coding (NTC), but some other papers use non-transformer architectures with residual connections and "simple" attention blocks and get comparable RD results lower runtime. For example:

1. [The Devil Is in the Details: Window-based Attention for Image Compression](https://arxiv.org/abs/2203.08450) -- they present a model very similar to swint-charm but also discuss a simpler windowed attention model with comparable performance.

2. [ELIC: Efficient Learned Image Compression with Unevenly Grouped Space-Channel Contextual Adaptive Coding](https://arxiv.org/abs/2203.10886) -- RD performance is on par with swint-charm with a faster model. I like their idea of uneven splits in CHARM as a way to better explore the RDC frontier. Mixing CHARM and a checkerboard decomposition is also a good idea, though I haven't seen much of an improvement over CHARM alone in my preliminary experiments (but that could be due to a missing implementation or training detail).

**Comment by Nikolai10** [comment]
Hello @minnend, thanks a lot for sharing your thoughts! You might find this interesting as well: https://github.com/Nikolai10/LIC-TCM.

I was also wondering if you are going to release the benchmarking code used in ["Advancing the Rate-Distortion-Computation Frontier for Neural Image Compression"](https://ieeexplore.ieee.org/document/10222381). I am quite curious how exactly the overall setup looks like (for PyTorch the go-to variant would be the [deepspeed profiler](https://www.deepspeed.ai/tutorials/flops-profiler/#usage-outside-the-deepspeed-runtime), is there a TensorFlow equivalent?)
